pkg/kotsadmsnapshot/backup.go

```diff
@@ -191,254 +201,512 @@ func CreateApplicationBackup(ctx context.Context, a *apptypes.App, isScheduled b
+		veleroBackup.Spec.IncludedNamespaces = append(veleroBackup.Spec.IncludedNamespaces, metadata.backupStorageLocationNamespace)
+	}
+
+	for _, appMeta := range metadata.apps {
+		// Don't merge the backup spec if we are using the new improved DR.
+		if !hasAppSpec {
+			err := mergeAppBackupSpec(veleroBackup, appMeta, metadata.kotsadmNamespace, metadata.ec != nil)
+			if err != nil {
+				return nil, errors.Wrap(err, "failed to merge app backup spec")
+			}
 		}
+	}
 
-		backupSpec, err := kotsKinds.Marshal("velero.io", "v1", "Backup")
+	veleroBackup.Annotations, err = appendCommonAnnotations(k8sClient, veleroBackup.Annotations, metadata, hasAppSpec)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to add annotations to backup")
+	}
+	if hasAppSpec {
+		veleroBackup.Annotations[InstanceBackupTypeAnnotation] = InstanceBackupTypeKotsadm
+	} else {
+		veleroBackup.Annotations[InstanceBackupTypeAnnotation] = InstanceBackupTypeCombined
+	}
+
+	if metadata.ec != nil {
+		veleroBackup.Spec.IncludedNamespaces = append(veleroBackup.Spec.IncludedNamespaces, ecIncludedNamespaces(metadata.ec.installation)...)
+	}
+
+	if metadata.snapshotTTL > 0 {
+		veleroBackup.Spec.TTL = metav1.Duration{
+			Duration: metadata.snapshotTTL,
+		}
+	}
+
+	veleroBackup.Spec.IncludedNamespaces = prepareIncludedNamespaces(veleroBackup.Spec.IncludedNamespaces)
+
+	return veleroBackup, nil
+}
+
+var EnableImprovedDR = false
+
+// getAppInstanceBackup returns a backup spec only if this is Embedded Cluster and the vendor has
+// defined both a backup and restore custom resource (improved DR).
+func getAppInstanceBackupSpec(k8sClient kubernetes.Interface, metadata instanceBackupMetadata) (*velerov1.Backup, error) {
+	// TODO(improveddr): remove this once we have fully implemented the improved DR
+	if !EnableImprovedDR {
+		return nil, nil
+	}
+
+	if metadata.ec == nil {
+		return nil, nil
+	}
+
+	var appVeleroBackup *velerov1.Backup
+
+	for _, appMeta := range metadata.apps {
+		// if there is both a backup and a restore spec this is using the new improved DR
+		if appMeta.kotsKinds.Backup == nil || appMeta.kotsKinds.Restore == nil {
+			continue
+		}
+
+		if len(metadata.apps) > 1 {
+			return nil, errors.New("cannot create backup for Embedded Cluster with multiple apps")
+		}
+
+		backupSpec, err := appMeta.kotsKinds.Marshal("velero.io", "v1", "Backup")
 		if err != nil {
 			return nil, errors.Wrap(err, "failed to get backup spec from kotskinds")
 		}
 
 		if backupSpec == "" {
-			continue
+			return nil, errors.New("backup spec is empty, this is unexpected")
 		}
 
-		appsSequences[a.Slug] = parentSequence
-		appVersions[a.Slug] = kotsKinds.Installation.Spec.VersionLabel
-
-		renderedBackup, err := helper.RenderAppFile(a, nil, []byte(backupSpec), kotsKinds, kotsadmNamespace)
+		renderedBackup, err := helper.RenderAppFile(appMeta.app, nil, []byte(backupSpec), appMeta.kotsKinds, metadata.kotsadmNamespace)
```

Comment:
we don't have to render the backup spec for EC as it only has newer versions of kots that store the rendered version kots kinds. so this can be removed.

Reply:
the marshal and parse parts can also be removed.

Reply:
done